GPU/Debugger/Record.cpp

```diff
@@ -0,0 +1,1013 @@
+﻿// Copyright (c) 2017- PPSSPP Project.
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, version 2.0 or later versions.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License 2.0 for more details.
+
+// A copy of the GPL 2.0 should have been included with the program.
+// If not, see http://www.gnu.org/licenses/
+
+// Official git repository and contact information can be found at
+// https://github.com/hrydgard/ppsspp and http://www.ppsspp.org/.
+
+#include <algorithm>
+#include <cstring>
+#include <vector>
+#include <snappy-c.h>
+#include "base/stringutil.h"
+#include "Common/Common.h"
+#include "Common/FileUtil.h"
+#include "Common/Log.h"
+#include "Core/Core.h"
+#include "Core/CoreTiming.h"
+#include "Core/ELF/ParamSFO.h"
+#include "Core/FileSystems/MetaFileSystem.h"
+#include "Core/HLE/sceDisplay.h"
+#include "Core/HLE/sceKernelMemory.h"
+#include "Core/MemMap.h"
+#include "Core/MIPS/MIPS.h"
+#include "Core/System.h"
+#include "GPU/GPUInterface.h"
+#include "GPU/GPUState.h"
+#include "GPU/ge_constants.h"
+#include "GPU/Common/TextureDecoder.h"
+#include "GPU/Common/VertexDecoderCommon.h"
+#include "GPU/Debugger/Record.h"
+
+namespace GPURecord {
+
+static const char *HEADER = "PPSSPPGE";
+static const int VERSION = 2;
+
+static bool active = false;
+static bool nextFrame = false;
+static bool writePending = false;
+
+enum class CommandType : u8 {
+	INIT = 0,
+	REGISTERS = 1,
+	VERTICES = 2,
+	INDICES = 3,
+	CLUT = 4,
+	TRANSFERSRC = 5,
+	MEMSET = 6,
+	MEMCPYDEST = 7,
+	MEMCPYDATA = 8,
+	DISPLAY = 9,
+
+	TEXTURE0 = 0x10,
+	TEXTURE1 = 0x11,
+	TEXTURE2 = 0x12,
+	TEXTURE3 = 0x13,
+	TEXTURE4 = 0x14,
+	TEXTURE5 = 0x15,
+	TEXTURE6 = 0x16,
+	TEXTURE7 = 0x17,
+};
+
+#pragma pack(push, 1)
+
+struct Command {
+	CommandType type;
+	u32 sz;
+	u32 ptr;
+};
+
+#pragma pack(pop)
+
+static std::vector<u8> pushbuf;
+static std::vector<Command> commands;
+static std::vector<u32> lastRegisters;
+static std::vector<u32> lastTextures;
+
+// TODO: Maybe move execute to another file?
+static u32 execMemcpyDest;
+static u32 execListBuf;
+static u32 execListPos;
+static u32 execListID;
+static const int LIST_BUF_SIZE = 256 * 1024;
+static std::vector<u32> execListQueue;
+
+class BufMapping {
```

Comment:
I think BufMapping needs some short high-level comments, just a quick overview of how it's used, and a quick explanation of the Slab system. Just really short, don't intend to create much extra work for you here.